### PR TITLE
Add metrics for AzureCloudDestination

### PR DESCRIPTION
**Pull request by lightningrob**
Add first part of uuid to Azure blob name. (Required public BlobId.getUuid())
Increase test coverage for CloudDataNode and CloudReplica

### CODE REVIEW

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 1]
```diff
@@ -14,6 +14,7 @@
 package com.github.ambry.cloud.azure;
 
 import com.codahale.metrics.MetricRegistry;
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: Not used.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 2]
```diff
@@ -267,6 +293,29 @@ private CloudBlobContainer getContainer(BlobId blobId, boolean autoCreate)
...
   }
 
+  /**
+   * Get the blob name to use in Azure Blob Storage
+   * @param blobId The {@link BlobId} to store.
+   * @return An Azure-friendly blob name.
+   */
+  private String getAzureBlobName(BlobId blobId) {
+    // Prefix to assist in blob data sharding, since beginning of blobId has little variation.
+    String prefix = blobId.getUuid().substring(0,4) + "-";
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: ```return blobId.getUuid().substring(0,4) + "-" + blobId.getID()```?

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestinationFactory.java [hunk 3]
```diff
@@ -29,16 +29,18 @@
   private static final Logger logger = LoggerFactory.getLogger(AzureCloudDestinationFactory.class);
   private final AzureCloudConfig azureCloudConfig;
   private final MetricRegistry metricRegistry;
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestinationFactory.java]: Can be a local var.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 4]
```diff
@@ -128,25 +130,36 @@ public boolean uploadBlob(BlobId blobId, long blobSize, CloudBlobMetadata cloudB
...
+      Timer.Context storageTimer = azureMetrics.blobUploadTime.time();
       CloudBlobContainer azureContainer = getContainer(blobId, true);
-      CloudBlockBlob azureBlob = azureContainer.getBlockBlobReference(blobId.getID());
+      String azureBlobName = getAzureBlobName(blobId);
+      CloudBlockBlob azureBlob = azureContainer.getBlockBlobReference(azureBlobName);
 
       if (azureBlob.exists()) {
         logger.debug("Skipping upload of blob {} as it already exists in Azure container {}.", blobId,
             azureContainer.getName());
+        azureMetrics.blobSkippedCount.inc();
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: Is it better to unify prefix of `blobSkippedCount `, `blobUploadedCount ` and `blobUploadErrorCount `?

`blobUploadSkipCount`, `blobUploadSuccessCount` and `blobUploadErrorCount`

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java (hunk 2 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: I prefer to make the prefix explicit and documented.

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java (hunk 4 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: Changed to blobUploadSkippedCount

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java [hunk 5]
```diff
@@ -267,6 +322,29 @@ private CloudBlobContainer getContainer(BlobId blobId, boolean autoCreate)
...
+  }
+
+  /**
+   * Update the appropriate error metrics corresponding to the thrown exception.
+   * @param e the exception thrown.
+   */
+  private void updateErrorMetrics(Exception e) {
+    if (e instanceof DocumentClientException) {
+      azureMetrics.documentErrorCount.inc();
+    } else {
```
Comment (zzmao) [inline ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]:
```
else if (e instanceof StorageException) {
}
```
To exclude  `URISyntaxException` explicitly.

File: ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureCloudDestinationTest.java [hunk 6]
```diff
@@ -193,25 +225,60 @@ public void testInitClientException() throws Exception {
     }
   }
 
+  /** Test setting connection policy with proxy from system properties */
+  @Test
+  public void testDocumentClientConnectionPolicy() throws Exception {
+    String proxyHost = "zwus2-proxy.azure.linkedin.com";
```
[the remark this answers is not in the document]
Comment (lightningrob) [reply ambry-cloud/src/test/java/com.github.ambry.cloud/azure/AzureCloudDestinationTest.java]: ok

File: ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java (hunk 5 above)
Comment (lightningrob) [reply ambry-cloud/src/main/java/com.github.ambry.cloud/azure/AzureCloudDestination.java]: azureContainer.getBlockBlobReference() can throw URISyntaxException, and this should count as a storageError. (Though unlikely to happen in practice.)

**Pull request merged**